contracts/linear/src/epoch_actions.rs

```diff
@@ -204,6 +204,137 @@ impl LiquidStakingContract {
             self.stake_amount_to_settle = 0;
         }
     }
+
+    /// Due to shares calculation and rounding of staking pool contract,
+    /// the amount of staked and unstaked balance might be a little bit
+    /// different than we requested.
+    /// This method is to sync the actual numbers with the validator.
+    pub fn sync_account_balance(&mut self, validator_id: AccountId) {
+        let min_gas = GAS_SYNC_BALANCE + GAS_EXT_GET_ACCOUNT + GAS_CB_VALIDATOR_SYNC_BALANCE;
+        require!(
+            env::prepaid_gas() >= min_gas,
+            format!("{}. require at least {:?}", ERR_NO_ENOUGH_GAS, min_gas)
+        );
+
+        let validator = self
+            .validator_pool
+            .get_validator(&validator_id)
+            .expect(ERR_VALIDATOR_NOT_EXIST);
+
+        validator
+            .sync_account_balance()
+            .then(ext_self_action_cb::validator_get_account_callback(
+                validator.account_id,
+                env::current_account_id(),
+                NO_DEPOSIT,
+                GAS_CB_VALIDATOR_SYNC_BALANCE
+            ));
+    }
+
+    /// This method is designed to drain a validator.
+    /// The weight of target validator should be set to 0 before calling this.
+    /// And a following call to drain_withdraw MUST be made after 4 epoches.
+    pub fn drain_unstake(&mut self, validator_id: AccountId) {
+        self.assert_manager();
+
+        // make sure enough gas was given
+        let min_gas = GAS_DRAIN_UNSTAKE + GAS_EXT_UNSTAKE + GAS_CB_VALIDATOR_UNSTAKED;
+        require!(
+            env::prepaid_gas() >= min_gas,
+            format!("{}. require at least {:?}", ERR_NO_ENOUGH_GAS, min_gas)
+        );
+
+        let mut validator = self.validator_pool.get_validator(&validator_id)
+            .expect(ERR_VALIDATOR_NOT_EXIST);
+
+        // make sure the validator:
+        // 1. has weight set to 0
+        // 2. not in pending release
+        // 3. has not unstaked balance (because this part is from user's unstake request)
+        require!(
+            validator.weight == 0,
+            ERR_NON_ZERO_WEIGHT
+        );
+        require!(
+            !validator.pending_release(),
+            ERR_VALIDATOR_UNSTAKE_WHEN_LOCKED
+        );
+        require!(
+            validator.unstaked_amount == 0,
```

Comment:
Is this necessary? if there's  non-zero `unstaked_amount`, the amount would be withdrawn in `drain_withdraw()` later.

Or if `unstaked_amount` is greater than `0`, should we run `drain_withdraw()` first?

Reply:
if `unstake_amount > 0` it means there are unstaked balance from users' unstake request, which should not be re-distributed later by `drain_withdraw`. If we don't make this assertion, then `drain_withdraw` will re-distribute those balance as well.

Reply:
I see.